src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt

```diff
@@ -0,0 +1,131 @@
+    var matchedPolicy: String? = null
+    var highestPriority: Int = -1
+    this.filter { (_, template) ->
+        template.lastUpdatedTime.toEpochMilli() < indexMetadata.creationDate
+    }.forEach { (policyID, template) ->
+        val matched = template.indexPatterns.stream().anyMatch(patternMatchPredicate)
+        if (matched && highestPriority < template.priority) {
+            highestPriority = template.priority
+            matchedPolicy = policyID
+        }
+    }
+
+    return matchedPolicy
+}
+
+/**
+ * validate the template Name and indexPattern provided in the template
+ * reusing ES validate function in MetadataIndexTemplateService
+ */
+@Suppress("ComplexMethod")
+fun validateFormat(indexPatterns: List<String>): ElasticsearchException? {
+    val validationErrors = mutableListOf<String>()
+    for (indexPattern in indexPatterns) {
+        if (indexPattern.contains(" ")) {
+            validationErrors.add("index_patterns [$indexPattern] must not contain a space")
+        }
+        if (indexPattern.contains(",")) {
+            validationErrors.add("index_pattern [$indexPattern] must not contain a ','")
+        }
+        if (indexPattern.contains("#")) {
+            validationErrors.add("index_pattern [$indexPattern] must not contain a '#'")
+        }
+        if (indexPattern.contains(":")) {
+            validationErrors.add("index_pattern [$indexPattern] must not contain a ':'")
+        }
+        if (indexPattern.startsWith("_")) {
+            validationErrors.add("index_pattern [$indexPattern] must not start with '_'")
+        }
+        if (!Strings.validFileNameExcludingAstrix(indexPattern)) {
+            validationErrors.add("index_pattern [" + indexPattern + "] must not contain the following characters " +
+                Strings.INVALID_FILENAME_CHARS)
+        }
+    }
+
+    if (validationErrors.size > 0) {
+        val validationException = ValidationException()
+        validationException.addValidationErrors(validationErrors)
+        return IndexManagementException.wrap(validationException)
+    }
+    return null
+}
+
+/**
+ * find policy templates whose index patterns overlap with given template
+ *
+ * @return map of overlapping template name to its index patterns
+ */
+@Suppress("SpreadOperator")
+fun Map<String, ISMTemplate>.findConflictingPolicyTemplates(
+    candidate: String,
+    indexPatterns: List<String>,
+    priority: Int
+): Map<String, List<String>> {
+    val automaton1 = Regex.simpleMatchToAutomaton(*indexPatterns.toTypedArray())
+    val overlappingTemplates = mutableMapOf<String, List<String>>()
+
+    // focus on template with same priority
+    this.filter { it.value.priority == priority }
+        .forEach { (policyID, template) ->
+        val automaton2 = Regex.simpleMatchToAutomaton(*template.indexPatterns.toTypedArray())
+        if (!Operations.isEmpty(Operations.intersection(automaton1, automaton2))) {
```

Comment:
Just to make sure, since this is an intersection of automatons, it will cover cases where one pattern is a subset of another right?

For example, say one template is created with pattern `exa*` and another is created with `exam*` both with the same priority. That case should considered a conflict.

Could we also add a test for this case. I see there is one for overlapping patterns when the pattern is the exact same but this particular case would be good to capture as well.

Reply:
changed that test to have 2 templates with index patterns `log*` and `lo*`